### PR TITLE
GPM sandcastle

### PR DESCRIPTION
**Pull request by javagl**
Edit: This has been updated to build upon https://github.com/CesiumGS/cesium/pull/12237 , which therefore should be merged first

# Description

Support for the `NGA_gpm_local` extension has recently been added to Cesium JS, via [#12204](https://github.com/CesiumGS/cesium/pull/12204). This (indirectly) includes the possibility to pick metadata values from the GPM PPE (Per-Point Error) textures, by treating them as property textures and apply [metadata picking](https://github.com/CesiumGS/cesium/pull/12075). 

This PR adds a Sandcastle that demonstrates the GPM visualization.

For now, this Sandcastle is in the `/Development` section, but can easily be moved elsewhere. It includes the option to visualize property texture values, and therefore might eventually replace https://sandcastle.cesium.com/index.html?src=Custom%20Shaders%20Property%20Textures.html 

The current state is a draft. It already replicates most of the functionality from an existing demo. But unfortunately, there seems to be an issue that is related to scaling and picking the metadata values from property textures for this data set: The picking functionality currently returns wrong values. The reason for this still has to be investigated.

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [ ] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [ ] I have updated the inline documentation, and included code examples where relevant
- [ ] I have performed a self-review of my code

### CODE REVIEW

**Comment by lilleyse** [comment]
> For now, this Sandcastle is in the `/Development` section, but can easily be moved elsewhere

I think the 3D Tiles folder would be better, more visible and easier to access from sandcastle.cesium.com.

**Comment by javagl** [comment]
I can move it there with the next update.

About the wrong metadata values:

This is related to the handling of `offset/scale`. The `ppeTexture` contains `offset/scale`. And there is no concept of `normalized` in the PPE Textures. But when translated to property textures, they **have** to be `normalized` in order to even be able to apply `offset/scale`. So in the property texture, `normalized` has to be set to `true`, and then taken into account in the `scale` factor. This was [anticipated in the GPM part](https://github.com/CesiumGS/cesium/pull/12204/files#diff-b245447cf491888173c8b66470aec6b9c1b9ea4a889aa7913e1b8577e7d85e0dR270). But on the metadata picking side, every `normalized/offset/scale` basically has to be "undone", to fully exploid the 256 bytes of the picking frame buffer. Right now, it only [took `normalized` into account](https://github.com/CesiumGS/cesium/pull/12075/files#diff-6709f4b7f058f59687308576a9f30f6657790bf5ebcac5d58a38467c1f0d4f30R480), but it essentially has to do what [`apply/unapplyValueTransform`](https://github.com/CesiumGS/cesium/blob/baaabaa49058067c855ad050be73a9cdfe9b6ac7/packages/engine/Source/Scene/MetadataComponentType.js#L360) are doing. Things might become more tricky for more complex cases, where the 'class property' defines some `offset/scale`, and the 'property texture property' _overrides_ the `offset/scale`. Care has to be taken to get the forward- and backward conversions right...

I do have a fix for that locally that "works" for the GPM data. But I'd like to test (and _describe_/document!) it more thoroughly.

**Comment by ggetz** [comment]
@javagl I've merged https://github.com/CesiumGS/cesium/pull/12237, and subsequently merged `main` into this PR. Is it ready to be moved out of Draft and be reviewed?

As an aside, it looks like this PR has two copies of the Sandcastle, one in `Developement/` and one at the gallery root. Is that intentional?

**Comment by javagl** [comment]
> As an aside, it looks like this PR has two copies of the Sandcastle, one in Developement/ and one at the gallery root. Is that intentional?

No, let me check...

**Comment by javagl** [comment]
Indeed, the one in `development` was the initial version - this should have been removed when adding the one in the (non-development) 3D Tiles section ( https://github.com/CesiumGS/cesium/pull/12232/commits/b4e5af441bce29cf58800c8e80c06b14477d59b1 ), but it was removed now.

File: Apps/Sandcastle/gallery/3D Tiles NGA GPM Visualization.html [hunk 1]
```diff
@@ -0,0 +1,1352 @@
...
+           * the selected tileset changes.
+           *
+           * @param {Cesium3DTileset|undefined} oldCurrentTileset The old tileset
+           * @param {Cesium3DTileset|undefined} newCurrentTileset The new tileset
+           */
+          updateForCurrentTilesetChanged(oldCurrentTileset, newCurrentTileset) {
+            if (oldCurrentTileset !== undefined) {
+              oldCurrentTileset.tileVisible.removeEventListener(this.tileVisibleListener);
+            }
+            if (newCurrentTileset !== undefined) {
```
Comment (ggetz) [inline Apps/Sandcastle/gallery/3D Tiles NGA GPM Visualization.html:1257]: Nitpick:

```suggestion
            if (Cesium.defined(oldCurrentTileset)) {
              oldCurrentTileset.tileVisible.removeEventListener(this.tileVisibleListener);
            }
            if (Cesium.defined(newCurrentTileset)) {
```

**Review by ggetz — COMMENTED** [review]
Overall this Sandcastle is looking and working great @javagl!

I have a few nitpick-y code comments.

I did notice that lower-resolution LODs of the tilesets, seen when zoomed out, weren't getting the Custom Shader applied, nor any ellipsoid anchor points, and returned "unknown" values when picked.

image

This _seems_ to be expected behavior, but let me know if it's not.

**Comment by javagl** [comment]
> This seems to be expected behavior, but let me know if it's not.

It is expected. There has been some discussion about how this level could be identified to begin with. When you're loading the initial tileset, you don't know how deeply you have to zoom to _have_ that information (i.e. you don't know in which level these things will be available). That's one reason why the initial view configurations for the sample data sets has been chosen carefully to be zoomed in closely enough. 

I also thought about how that could be made more apparent. Indicating it _visually_ could be difficult. There might not be a silver bullet in general. But maybe mentioning this in the ["info box" text](https://github.com/CesiumGS/cesium/pull/12232/files#diff-d344962f2d24b7e2aecea8f2641a2c5365f0acdc645bcafed180e3f587a7fb8bR107) might be one way...?

**Comment by ggetz** [comment]
Thanks @javagl! I mostly just wanted to confirm from a testing testing standpoint. 

At least at this point, I don't think further action is needed to indicate this.

**Review by ggetz — APPROVED** [review]

**Pull request merged**